origin-growth/src/rules/rules.js

```diff
@@ -0,0 +1,350 @@
+const Sequelize = require('sequelize')
+
+const db = require('../models')
+const { GrowthEventTypes, GrowthEventStatuses } = require('../enums')
+
+// System cap for number of rewards per rule.
+const MAX_NUM_REWARDS_PER_RULE = 1000
+
+
+class Reward {
+  constructor(campaignId, levelId, ruleId, money) {
+    this.campaignId = campaignId
+    this.levelId = levelId
+    this.ruleId = ruleId
+    this.money = money
+  }
+}
+
+
+class Campaign {
+  constructor(campaign, config) {
+    this.campaign = campaign
+    this.config = config
+
+    if (!this.config.numLevels ||
+      !Number.isInteger(this.config.numLevels) ||
+      this.config.numLevels <= 0) {
+      throw new Error(`Campaign ${campaign.id}: invalid or missing numLevels field.`)
+    }
+    this.numLevels = this.config.numLevels
+
+    this.levels = {}
+    for (let i = 0; i < this.config.numLevels; i++) {
+      if (!this.config.levels[i]) {
+        throw new Error(`Campaign ${this.campaign.id}: missing level ${i}`)
+      }
+      this.levels[i] = new Level(this.campaign.id, i, this.config.levels[i])
+    }
+  }
+
+  /**
+   * Reads events related to a user from the DB.
+   * @param {string} ethAddress - User's account.
+   * @param {boolean} duringCampaign - Restricts query to events that occurred
+   *  during the campaign vs since user signed up.
+   * @returns {Promise<List<models.GrowthEvent>>}
+   */
+  async getEvents(ethAddress, duringCampaign) {
+    const whereClause = {
+      ethAddress: ethAddress.toLowerCase()
+    }
+    if (duringCampaign) {
+      whereClause.createdAt = {
+        [Sequelize.Op.gte]: this.campaign.startDate,
+        [Sequelize.Op.lt]: this.campaign.endDate
+      }
+    }
+    const events = await db.GrowthEvent.findAll({
+        where: whereClause,
+        order: [ ['id', 'ASC'] ],
+      }
+    )
+    return events
+  }
+
+  /**
+   * Calculates the current campaign level the user is at.
+   * Considers events that occurred since user joined the platform.
+   *
+   * @param {string} ethAddress - User's account.
+   * @returns {Promise<number>}
+   */
+  async getCurrentLevel(ethAddress) {
+    const events = await this.getEvents(ethAddress, false)
+    let level
+    for (level = 0; level < this.config.numLevels - 1; level++) {
+      if (!this.levels[level].qualifyForNextLevel(ethAddress, events)) {
+        break
+      }
+    }
+    return level
+  }
+
+  /**
+   * Calculates rewards earned by the user.
+   * Only considers events that occurred during the campaign.
+   *
+   * @param {string} ethAddress - User's account.
+   * @returns {Promise<Object>} - Dictionary with rewards grouped first by level and second by rule.
```

Comment:
@sparrowDom Let me know if this grouping makes sense to you. I thought grouping by level and by rule would make it easier for the UI to show rewards per level/rule. But I'm totally open to suggestion. For example, if that's easier for you, we could simply return a flat list of Reward objects.

Reply:
Thanks for asking :) Both are good, but if I can choose I would prefer the flat list.

Reply:
Done